### ISSUE TITLE
Why DOTNET_GCHeapHardLimitPercent not working?

### ISSUE THREAD

**Issue by careless6666** [dotnet/runtime#92113]
Hi. We have dotnet application with  <TargetFramework>net7.0</TargetFramework>
and k8s limits
```
resources:
  limits:
    cpu: "4"
    memory: "3Gi"
  requests:
    cpu: "1.5"
    memory: "2Gi"
```

and dotnet variables
```
env:
  - name: ASPNETCORE_ENVIRONMENT
    value: "Production"
  - name: DOTNET_gcServer
    value: "1"
  - name: DOTNET_gcConcurrent
    value: "1"
  - name: DOTNET_GCConserveMemory
    value: "5"
  - name: DOTNET_GCRetainVM
    value: "0"
  - name: DOTNET_GCHeapHardLimitPercent
    value: "0x28"
```

My metrics from gc
<img width="1663" alt="Screenshot 2023-09-15 at 1 23 28 PM" src="https://github.com/dotnet/runtime/assets/8223175/bbdaafaa-fe0a-4251-ba0e-00c9c570b535">
dotnet_totalmemory
<img width="1648" alt="Screenshot 2023-09-15 at 1 23 49 PM" src="https://github.com/dotnet/runtime/assets/8223175/f359664d-76d3-4b74-ae16-2e86a170473b">
dotnet_collection_count_total

And metrics from os
<img width="1653" alt="Screenshot 2023-09-15 at 1 24 23 PM" src="https://github.com/dotnet/runtime/assets/8223175/ecd19dd0-c789-48bb-9d31-6758475ce6d9">
I set up DOTNET_GCRetainVM and expected see same memory allocation for dotnet and for os memory. 
So going beyond the limits of memory usage would be ok if it were short-term, but here it is clearly not going to go to the set 40%
We research this question, because we receive oom kill, on some applications, and there is no memory leak in these applications. If you call GC.Collect() memoey will be cleaned

I expect the memory to stay at 1.2 GB (40%)

**Comment by janvorli** [comment]
@careless6666 the `DOTNET_GCHeapHardLimitPercent` should be a hexadecimal value, but without the `0x` prefix. All DOTNET_xxx numeric values work this way.

**Comment by careless6666** [comment]
I changed values to this

```
env:
  - name: ASPNETCORE_ENVIRONMENT
    value: "Production"
  - name: DOTNET_gcServer
    value: "1"
  - name: DOTNET_gcConcurrent
    value: "1"
  - name: DOTNET_GCConserveMemory
    value: "5"
  - name: DOTNET_GCRetainVM
    value: "0"
  - name: DOTNET_GCHeapHardLimitPercent
    value: "28"
```
but result not changed
<img width="1659" alt="Screenshot 2023-09-20 at 9 58 40 PM" src="https://github.com/dotnet/runtime/assets/8223175/57e3aa16-b73b-4803-a96e-ffaccd8ea9e5">

**Comment by mangod9** [comment]
can you please provide a sample repro for this?

**Comment by careless6666** [comment]
i try to do it

**Comment by careless6666** [comment]
Now I'm reproducing another exception in this service, but I think I can add it to this issue because it's also a gc settings issue.
https://github.com/careless6666/gc-issue

I call 2 endpoint on this service 
init  - one call
```
curl -X 'POST' \
  'http://playground-dotnet-latency-issue-feature-gc-noplatform.my-company.net/fake-work/init' \
```  
and 100 rps to test endpoint
```
curl -X 'GET' \
  'http://playground-dotnet-latency-issue-feature-gc-noplatform.my-company.net/fake-work/user' \
```
we have 2 gb memory limit, 40% hard limit for heap => 800mb. I check again documentation and didn't find what's have to happened when limit will be achieved. I expected long pause and slow response for my queries because I expected that gc will start  for clean memory, but i receive out of memory exception. 

<img width="826" alt="Screenshot 2023-10-04 at 10 38 48 AM" src="https://github.com/dotnet/runtime/assets/8223175/429662b1-d70a-4d97-be0b-6d90c2d21670">

**Comment by janvorli** [comment]
Thank you, I'll give it a try.

**Comment by janvorli** [comment]
@careless6666 this testing app seems to have a real managed leak. I have attached lldb to the process after it started ooming and the GC heap size was consuming almost all of the allocated space.
As for the percentage limit, the oom happens at about 860MB for me with your app (the runtime allocates some native memory too, that's why it is higher than the GC limit). I was running it just with a docker, but I believe it should behave the same way in Kubernetes.

I've dumped the GC heap stats at the time of the timeout, here are the objects that allocated the largest amount of memory and their count (the size is summary for each type):
```
7fe9e679dd40   1,132     117,728 System.Reflection.RuntimeMethodInfo
7fe9e59b8690     198     746,640 System.Int32[]
7fe9e6806618   2,077   1,999,335 System.Byte[]
7fe9ea04adf8       1   3,754,512 System.Collections.Generic.Dictionary<System.Int64, Playground.Service.LatencyIssue.BLL.Services.User>+Entry[]
7fe9e90c3180 100,001   4,000,040 Playground.Service.LatencyIssue.BLL.Services.AdditionalInfo
7fe9e90c3060 100,001   4,000,040 Playground.Service.LatencyIssue.BLL.Services.User
55ba17bd5f90  11,446  21,253,040 Free
7fe9e59bd478 204,732 729,630,494 System.String
Total 441,519 objects, 767,163,008 bytes
```

I have a loop that sends 10000 requests, I am not sure though after how many of them the oom occured. As you can see, there are ~100000 objects of the  Playground.Service.LatencyIssue.BLL.Services.AdditionalInfo and  the same number of Playground.Service.LatencyIssue.BLL.Services.User.

This is a relevant part of the object chain that keeps them alive:
```
          -> 7fe8f7813178     Microsoft.Extensions.DependencyInjection.ServiceLookup.ConstructorCallSite
          -> 7fe8f78132d8     Playground.Service.LatencyIssue.BLL.Services.FakeWorkService
          -> 7fe8f7813308     System.Collections.Generic.Dictionary<System.Int64, Playground.Service.LatencyIssue.BLL.Services.User>
          -> 7fe904000048     System.Collections.Generic.Dictionary<System.Int64, Playground.Service.LatencyIssue.BLL.Services.User>+Entry[]
          -> 7fe8fb700a20     Playground.Service.LatencyIssue.BLL.Services.User
          -> 7fe8fb700ab8     Playground.Service.LatencyIssue.BLL.Services.AdditionalInfo
```

**Comment by careless6666** [comment]
I try again my test and make several measures

My counters before start test

![Screenshot from 2023-10-05 23-09-52](https://github.com/dotnet/runtime/assets/8223175/a6b52d17-535d-429a-9374-62e8ab4a5eba)

after part of queries

![Screenshot from 2023-10-05 23-13-47](https://github.com/dotnet/runtime/assets/8223175/3fcacf31-5ac3-4793-be2c-2e4dbc02d27b)
Working set increased

I check my heap statistic

```
7fde3af2dd28     207      31,464 System.RuntimeType+RuntimeTypeCache
7fde3a7a13d8     883      35,320 System.Threading.ExecutionContext
7fde3b769818     737      35,376 System.Text.StringBuilder
7fde3aff58e8     466      37,280 System.Signature
7fde3a14e158     818      37,408 System.String[]
7fde3e3f8ac8      83      37,848 Interop+Sys+IOVector[]
7fde3e344c98     294      42,336 System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Threading.Tasks.VoidTaskResult>+AsyncStateMachineBox<Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker+<<InvokeNextActionFilterAsync>g__Awaited|10_0>d>
7fde3e3f35a8     147      45,864 System.Text.Json.WriteStackFrame[]
7fde3aff5538     533      51,168 System.Reflection.RuntimeParameterInfo
7fde39ff8390   1,504      60,160 System.RuntimeType
7fde3af2f7d8     784      81,536 System.Reflection.RuntimeMethodInfo
7fde3a1766c8     771      99,290 System.Char[]
7fde39ffa508     956     163,984 System.Object[]
7fde3a06b9d8     440     748,352 System.Int32[]
7fde3adbd4e0     582   2,316,721 System.Byte[]
7fde3e170088       1   3,754,512 System.Collections.Generic.Dictionary<System.Int64, Playground.Service.LatencyIssue.BLL.Services.User>+Entry[]
7fde3d53cc58 100,001   4,000,040 Playground.Service.LatencyIssue.BLL.Services.User
7fde3d53cd98 100,001   4,000,040 Playground.Service.LatencyIssue.BLL.Services.AdditionalInfo
564b39bbcd50   2,316   8,280,736 Free
7fde3a121038 210,366 108,698,114 System.String
Total 448,302 objects, 134,419,390 bytes
```

Yes, we see a lot of objects type of User, AdditionalInfo and Strings

when i make more queries and get againg dump and counters

![Screenshot from 2023-10-05 23-33-24](https://github.com/dotnet/runtime/assets/8223175/535ed06f-84ce-4cd0-82e7-707aebbd7e30)

check again heap statistic

```
7fde3e344518     823     138,264 System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Threading.Tasks.VoidTaskResult>+AsyncStateMachineBox<Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker+<<InvokeActionMethodAsync>g__Awaited|12_0>d>
7fde3de15a38   1,343     139,672 System.Diagnostics.Tracing.CounterPayload
7fde3e347970     823     144,848 System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Threading.Tasks.VoidTaskResult>+AsyncStateMachineBox<Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker+<<InvokeAsync>g__Logged|17_1>d>
7fde3a14e158   3,522     145,568 System.String[]
7fde3e343a80     823     151,432 System.Runtime.CompilerServices.AsyncTaskMethodBuilder<Microsoft.AspNetCore.Mvc.IActionResult>+AsyncStateMachineBox<Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor+AwaitableObjectResultExecutor+<Execute>d__0>
7fde3e3fa2e0     429     154,440 System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Threading.Tasks.VoidTaskResult>+AsyncStateMachineBox<System.Text.Json.JsonSerializer+<WriteStreamAsync>d__121<System.Object>>
7fde3ded7938   3,291     157,968 System.Threading.AsyncLocalValueMap+TwoElementAsyncLocalValueMap
7fde3d5340c8     823     158,016 Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker
7fde3ded3cd8     823     164,600 System.Diagnostics.Activity
7fde3e3f8ac8     429     195,624 Interop+Sys+IOVector[]
7fde3a7a13d8   4,940     197,600 System.Threading.ExecutionContext
7fde3b769818   4,176     200,448 System.Text.StringBuilder
7fde3e344c98   1,646     237,024 System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Threading.Tasks.VoidTaskResult>+AsyncStateMachineBox<Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker+<<InvokeNextActionFilterAsync>g__Awaited|10_0>d>
7fde3e3f35a8     823     256,776 System.Text.Json.WriteStackFrame[]
7fde39ffa508   4,406     323,656 System.Object[]
7fde3a1766c8   4,249     396,718 System.Char[]
7fde3a06b9d8   1,801     797,876 System.Int32[]
7fde3e170088       1   3,754,512 System.Collections.Generic.Dictionary<System.Int64, Playground.Service.LatencyIssue.BLL.Services.User>+Entry[]
7fde3d53cc58 100,001   4,000,040 Playground.Service.LatencyIssue.BLL.Services.User
7fde3d53cd98 100,001   4,000,040 Playground.Service.LatencyIssue.BLL.Services.AdditionalInfo
7fde3adbd4e0   2,445   5,267,446 System.Byte[]
564b39bbcd50   4,221   7,605,416 Free
7fde3a121038 241,909 189,045,398 System.String
```

We see that size of instances of User and AdditionalInfo not changed, but increased amount of strings

if we try to look at live strings

dumpheap -strings -live -min 100

...
 7f9e3fe767a8     7fde3a121038            118 
    7f9e3fe76848     7fde3a121038            110 
    7f9e3fe768e0     7fde3a121038            118 
    7f9e3fe76980     7fde3a121038            110 
    7f9e3fe76a18     7fde3a121038            118 
    7f9e3fe76ab8     7fde3a121038            110 

    7f9e3fe77e38     7fde3a121038            110 
    7f9e3fe77ed0     7fde3a121038            118 
    7f9e3fe77f70     7fde3a121038            110 
    7f9e3fe78008     7fde3a121038            118 
    7f9e3fe780a8     7fde3a121038            110 
    7f9e3fe78140     7fde3a121038            118 
    7f9e3fe781e0     7fde3a121038            110 
    7f9e3fe78278     7fde3a121038            118 
    7f9e3fe78318     7fde3a121038            110 
    7f9e3fe783b0     7fde3a121038            118 
    7f9e3fe78450     7fde3a121038            110 
...

We see a lot live strings which must be died 

dumpobj 7f9e3cd8d5a8    
```                                                                                                                                                                                 
Name:        System.String
MethodTable: 00007fde3a121038
EEClass:     00007fde3a076fc8
Tracked Type: false
Size:        118(0x76) bytes
File:        /usr/lib/dotnet/shared/Microsoft.NETCore.App/7.0.9/System.Private.CoreLib.dll
String:      fake_status_e1c4f554-3fd5-4c7a-ab91-bb1db7fe03ba
Fields:
              MT    Field   Offset                 Type VT     Attr            Value Name
00007fde39ffe378  40002b9        8         System.Int32  1 instance               48 _stringLength
00007fde3a025e48  40002ba        c          System.Char  1 instance               66 _firstChar
00007fde3a121038  40002b8       d8        System.String  0   static 00007f9e3f800218 Empty
```
And what interesting they don't have roots
```
> gcroot 7f9e3cd8d5a8                                                                                                                                                                                      
Found 0 unique roots.
```

I think that gc simply does not have time to mark and clean lines and in the end it just crashes. The problem is the frequent allocation of large strings

**Comment by careless6666** [comment]
We can try filter our live big strings, (I fix message for fake long strings)

dumpheap -type String -live -min 500

 ```
 7fb382023880     7ff383531038        216,076 
    7fb3820584b0     7ff383531038        216,076 
    7fb38208d0e0     7ff383531038        216,076 
    7fb3820c1d10     7ff383531038        216,076 
    7fb3820f6940     7ff383531038        216,076 
    7fb38212b570     7ff383531038        216,076 
    7fb3821601a0     7ff383531038        216,076 
    7fb382194dd0     7ff383531038        216,076 
    7fb3821c9a00     7ff383531038        216,076 
    7fb3821fe630     7ff383531038        216,076 
    7fb382233260     7ff383531038        216,076 
    7fb382267e90     7ff383531038        216,076 
    7fb38229cac0     7ff383531038        216,076 
    7fb3822d16f0     7ff383531038        216,076 
    7fb382306320     7ff383531038        216,076 
    7fb38233af50     7ff383531038        216,076 
    7fb38236fb80     7ff383531038        216,076 
    7fb3823a47b0     7ff383531038        216,076 
    7fb3823d93e0     7ff383531038        216,076 
    7fb38240e010     7ff383531038        216,076 
    7fb382442c40     7ff383531038        216,076 
    7fb382477870     7ff383531038        216,076 
    7fb3824ac4a0     7ff383531038        216,076 
    7fb3824e10d0     7ff383531038        216,076 
```

And look content of our string

```
dc -c 1024  7fb382023880                                                                                                                                                                                 
00007fb382023880: 1038 8353 7ff3 0000 a5fb 0001 0046 0061  ......Fa
00007fb382023890: 006b 0065 0020 006c 006f 006e 0067 0061  ke longa
00007fb3820238a0: 0035 0032 0065 0035 0032 0064 0031 002d  52e52d1-
00007fb3820238b0: 0065 0063 0061 0035 002d 0034 0031 0030  eca5-410
00007fb3820238c0: 0066 002d 0062 0066 0039 0061 002d 0033  f-bf9a-3
00007fb3820238d0: 0031 0030 0066 0031 0033 0034 0065 0035  10f134e5
00007fb3820238e0: 0064 0039 0030 0066 0032 0037 0036 0061  d90f276a
00007fb3820238f0: 0062 0065 0065 002d 0062 0031 0035 0035  bee-b155
00007fb382023900: 002d 0034 0066 0065 0033 002d 0038 0061  -4fe3-8a

```

Fake longa... it is our string which we concat and send in responses and never store in nowhere, this string must be died but it is still live

**Comment by janvorli** [comment]
Ah, you are right that the strings are the biggest offender. Somehow my eyes have filtered that entry out of the list probalby due to the fact they were after the Free.

However, I doubt that majority of those would have no root. Obviously, there will be some strings without roots, but it is very unlikely that GC would not collect hundreds of megabytes of those. I'll try to create some script in lldb to figure out how many of those have no root. 

I've tried to get gcroots of a couple of the strings you've mentioned and the ones I've tried were all rooted, directly referenced by the Playground.Service.LatencyIssue.BLL.Services.AdditionalInfo
